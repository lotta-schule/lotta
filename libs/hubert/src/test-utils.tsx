import * as React from 'react';
import { DefaultThemes } from '@lotta-schule/theme';
import { render, type RenderOptions } from '@testing-library/react';
import { MotionConfig } from 'framer-motion';
import { HubertProvider } from './HubertProvider';
import { GlobalStyles } from './theme';
import { BrowserNode, BrowserState, NodeList } from 'browser';
import { BrowserStateContext } from 'browser/BrowserStateContext';
import * as fixtures from 'test-fixtures';

const theme = DefaultThemes.standard;

const Wrapper = ({ children }: any) => (
  <HubertProvider>
    <MotionConfig transition={{ duration: 0 }}>
      <GlobalStyles theme={theme} />
      {children}
    </MotionConfig>
  </HubertProvider>
);

const customRender = (
  ui: React.ReactElement<any>,
  renderOptions: RenderOptions = {}
): ReturnType<typeof render> =>
  render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };

export const createPromise = <T = void,>() => {
  let resolve: (value: T) => void;
  let reject: (error?: any) => void;
  const promise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return { promise, resolve: resolve!, reject: reject! };
};

// For Testing components that use the BrowserStateContext
export type TestBrowserWrapperProps = {
  children?: React.ReactNode;
  defaultNodes?: BrowserNode[];
} & Partial<Omit<BrowserState, 'onRequestChildNodes' | 'renderNodeList'>>;
export const TestBrowserWrapper = ({
  children,
  mode = 'view-and-edit',
  defaultNodes = fixtures.browserNodes,
  moveNode = vi.fn(),
  renameNode = vi.fn(),
  deleteNode = vi.fn(),
  searchNodes = vi.fn(),
  currentPath = [],
  selected = [],
  isNodeDisabled = vi.fn(() => false),
  onSelect = vi.fn(),
  onNavigate = vi.fn(),
  setCurrentAction = vi.fn(),
  currentAction = null,
  currentSearchResults = null,
  setCurrentSearchResults = vi.fn(),
  canEdit,
  uploadClient = {
    addFile: vi.fn(),
    byState: { uploading: [], pending: [], done: [], error: [] },
    hasErrors: false,
    isSuccess: false,
    isUploading: false,
    currentProgress: null,
    currentUploads: [],
  },
}: TestBrowserWrapperProps) => {
  const [nodes, setNodes] = React.useState<BrowserNode[]>(defaultNodes);
  const [isFilePreviewVisible, setIsFilePreviewVisible] = React.useState(false);

  const onRequestChildNodes = (node: BrowserNode<'directory'> | null) =>
    nodes.filter((n) => n.parent === (node?.id ?? null));

  return (
    <BrowserStateContext.Provider
      value={{
        uploadClient,
        canEdit,
        onRequestChildNodes: async (node) =>
          onRequestChildNodes(node as BrowserNode<'directory'>),
        renderNodeList: ({ path }) => (
          <NodeList path={path} nodes={onRequestChildNodes(path.at(-1)!)} />
        ),
        currentAction,
        onSelect,
        onNavigate,
        currentSearchResults,
        currentPath,
        selected,
        mode,
        isNodeDisabled,
        setCurrentAction,
        setCurrentSearchResults,
        moveNode,
        renameNode,
        deleteNode,
        searchNodes,
        isFilePreviewVisible,
        setIsFilePreviewVisible,
        createDirectory: async (parentNode, name) => {
          setNodes((nodes) => {
            return [
              ...nodes,
              {
                id: `${nodes.length + 1}`,
                name,
                type: 'directory',
                parent: parentNode?.id ?? null,
                meta: {},
              },
            ];
          });
        },
      }}
    >
      {children}
    </BrowserStateContext.Provider>
  );
};

export * as fixtures from './test-fixtures';
