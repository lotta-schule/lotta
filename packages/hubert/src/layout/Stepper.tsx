import * as React from 'react';
import { KeyboardArrowLeft, KeyboardArrowRight } from '../icon';
import { NavigationButton } from '../button';
import clsx from 'clsx';

import styles from './Stepper.module.scss';

export type StepperProps = React.HTMLAttributes<HTMLDivElement> & {
  className?: string;
  style?: React.CSSProperties;
  maxSteps: number;
  currentStep: number;
  onStep: (_newStep: number) => void;
};

export const Stepper = React.memo(
  ({
    className,
    currentStep,
    maxSteps,
    style,
    onStep,
    ...props
  }: StepperProps) => {
    const ariaAttrs: React.AriaAttributes = {
      'aria-valuemin': 1,
      'aria-valuemax': maxSteps,
      'aria-valuenow': currentStep + 1,
    };
    return (
      <div
        role="spinbutton"
        {...ariaAttrs}
        className={clsx(styles.root, className)}
        style={style}
        {...props}
      >
        <NavigationButton
          small
          icon={<KeyboardArrowLeft />}
          onClick={() => onStep(currentStep - 1)}
          disabled={currentStep <= 0}
          aria-label={'Vorheriger Schritt'}
          style={{ color: 'rgba(var(--lotta-text-color), 1)' }}
        >
          vorheriges
        </NavigationButton>
        <div>
          {currentStep + 1} / {maxSteps}
        </div>
        <NavigationButton
          small
          icon={<KeyboardArrowRight />}
          onClick={() => onStep(currentStep + 1)}
          disabled={currentStep >= maxSteps - 1}
          aria-label={'Nächster Schritt'}
          style={{ color: 'rgba(var(--lotta-text-color), 1)' }}
        >
          nächstes
        </NavigationButton>
      </div>
    );
  }
);
Stepper.displayName = 'Stepper';
