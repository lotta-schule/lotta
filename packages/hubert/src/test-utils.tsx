import * as React from 'react';
import { DefaultThemes } from '@lotta-schule/theme';
import { render, RenderOptions } from '@testing-library/react';
import { MotionConfig } from 'framer-motion';
import { HubertProvider } from './HubertProvider';
import { GlobalStyles } from './theme';

const theme = DefaultThemes.standard;

const Wrapper = ({ children }: any) => (
  <HubertProvider>
    <MotionConfig transition={{ duration: 0 }}>
      <GlobalStyles theme={theme} />
      {children}
    </MotionConfig>
  </HubertProvider>
);

const customRender = (
  ui: React.ReactElement,
  renderOptions: Omit<RenderOptions, 'wrapper'> = {}
) =>
  render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
