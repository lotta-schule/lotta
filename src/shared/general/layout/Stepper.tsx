import * as React from 'react';
import { KeyboardArrowLeft, KeyboardArrowRight } from '@material-ui/icons';
import { NavigationButton } from '../button/NavigationButton';
import clsx from 'clsx';

import styles from './Stepper.module.scss';

export interface StepperProps {
    className?: string;
    style?: React.CSSProperties;
    maxSteps: number;
    currentStep: number;
    onStep: (newStep: number) => void;
}

export const Stepper = React.memo<StepperProps>(
    ({ className, currentStep, maxSteps, style, onStep }) => {
        return (
            <div className={clsx(styles.root, className)} style={style}>
                <NavigationButton
                    small
                    icon={<KeyboardArrowLeft />}
                    onClick={() => onStep(currentStep - 1)}
                    disabled={currentStep <= 0}
                    aria-label={'Vorheriger Schritt'}
                >
                    vorheriges
                </NavigationButton>
                <div>
                    {currentStep + 1} / {maxSteps}
                </div>
                <NavigationButton
                    small
                    icon={<KeyboardArrowRight />}
                    onClick={() => onStep(currentStep + 1)}
                    disabled={currentStep >= maxSteps - 1}
                    aria-label={'Nächster Schritt'}
                >
                    nächstes
                </NavigationButton>
            </div>
        );
    }
);
Stepper.displayName = 'Stepper';
