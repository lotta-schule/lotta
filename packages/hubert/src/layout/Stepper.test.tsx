import * as React from 'react';
import { render } from '../test-utils';
import { Stepper } from './Stepper';
import userEvent from '@testing-library/user-event';

describe('Stepper', () => {
  it('should render a stepper', () => {
    const screen = render(
      <Stepper currentStep={0} maxSteps={3} onStep={jest.fn()} />
    );
    expect(screen.getByRole('spinbutton')).toMatchSnapshot();
  });

  it('should render the correct step count', () => {
    const screen = render(
      <Stepper currentStep={0} maxSteps={3} onStep={jest.fn()} />
    );
    expect(screen.getByText('1 / 3')).toBeVisible();
  });

  describe('previous step', () => {
    it('should select the previous step on button click', async () => {
      const fireEvent = userEvent.setup();
      const onStep = jest.fn();
      const screen = render(
        <Stepper currentStep={1} maxSteps={3} onStep={onStep} />
      );
      await fireEvent.click(screen.getByRole('button', { name: /vorherig/i }));
      expect(onStep).toHaveBeenCalledWith(0);
    });

    it('should disable previous button when on first step', () => {
      const screen = render(
        <Stepper currentStep={0} maxSteps={3} onStep={jest.fn()} />
      );
      expect(screen.getByRole('button', { name: /vorherig/i })).toBeDisabled();
    });
  });

  describe('next step', () => {
    it('should select the next step on button click', async () => {
      const fireEvent = userEvent.setup();
      const onStep = jest.fn();
      const screen = render(
        <Stepper currentStep={1} maxSteps={3} onStep={onStep} />
      );
      await fireEvent.click(screen.getByRole('button', { name: /nächst/i }));
      expect(onStep).toHaveBeenCalledWith(2);
    });

    it('should disable previous button when on first step', () => {
      const screen = render(
        <Stepper currentStep={2} maxSteps={3} onStep={jest.fn()} />
      );
      expect(screen.getByRole('button', { name: /nächst/i })).toBeDisabled();
    });
  });
});
