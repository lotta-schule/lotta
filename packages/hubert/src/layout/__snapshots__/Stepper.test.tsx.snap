// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stepper should render a stepper 1`] = `
<div
  aria-valuemax="3"
  aria-valuemin="1"
  aria-valuenow="1"
  class="root"
  role="spinbutton"
>
  <button
    aria-current="false"
    aria-label="Vorheriger Schritt"
    class="root root root small"
    data-variant="default"
    disabled=""
    role="button"
    type="button"
  >
    <span
      class="icon icon"
    >
      <svg
        aria-hidden="true"
        class="root"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
        />
      </svg>
    </span>
    <span
      class="label label"
    >
      vorheriges
    </span>
  </button>
  <div>
    1
     / 
    3
  </div>
  <button
    aria-current="false"
    aria-label="Nächster Schritt"
    class="root root root small"
    data-variant="default"
    role="button"
    type="button"
  >
    <span
      class="icon icon"
    >
      <svg
        aria-hidden="true"
        class="root"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
        />
      </svg>
    </span>
    <span
      class="label label"
    >
      nächstes
    </span>
  </button>
</div>
`;
